File: rendering_api/metal/src/core/sync_objects/metal_semaphore.cpp
#include <core/sync_objects/gryphn_sync_semaphore.h>
#include "metal_semaphore.h"

GN_EXPORT gnReturnCode gnCreateSyncSemaphoreFn(gnSyncSemaphore* semaphore, const gnOutputDevice& device) {
    semaphore->semaphore = new gnPlatformSyncSemaphore();
    semaphore->semaphore->semaphore = dispatch_semaphore_create(0);
    // semaphore->semaphore->semaphore = device.outputDevice->device->newFence();
    return GN_SUCCESS;
}
GN_EXPORT void gnDestroySyncSemaphoreFn(const gnSyncSemaphore& semaphore) {
    // semaphore.semaphore->semaphore->release();
}

File: src/apis/metal/src/sync/semaphore/metal_semaphore.m
#include "metal_semaphore.h"
#include "core/devices/metal_output_devices.h"

gnReturnCode gnCreateSemaphoreFn(struct gnSemaphore_t* semaphore, struct gnOutputDevice_t* device) {
    semaphore->semaphore = malloc(sizeof(gnPlatformSemaphore));
    semaphore->semaphore->event = [device->outputDevice->device newEvent];

    return GN_SUCCESS;
}
void gnDestroySemaphoreFn(struct gnSemaphore_t* semaphore) {
    [semaphore->semaphore->event release];
    free(semaphore->semaphore);
}
